src/neural/loader.cc

```
@@ -46,17 +50,81 @@ FloatVectors LoadFloatsFromFile(const std::string& filename) {
     } else {
       bytes_read += sz;
       buffer.resize(bytes_read);
-      // Add newline in the end for the case it was not there.
-      buffer.push_back('\n');
       break;
     }
   }
   gzclose(file);
 
+  return buffer;
+}
+
+void DenormLayer(const lc0::Weights_Layer& layer, FloatVectors& vecs) {
+  FloatVector v;
+  auto& buffer = layer.params();
+  auto data = reinterpret_cast<const std::uint16_t*>(buffer.data());
+  int n = buffer.length() / 2;
+  v.resize(n);
+  for (int i = 0; i < n; i++) {
+    v[i] = data[i] / float(0xffff);
+    v[i] *= layer.max_val() - layer.min_val();
+    v[i] += layer.min_val();
+  }
+  vecs.emplace_back(v);
+}
+
+void DenormConvBlock(const lc0::Weights_ConvBlock& conv, FloatVectors& vecs) {
+  DenormLayer(conv.bn_stddivs(), vecs);
+  DenormLayer(conv.bn_means(), vecs);
+  DenormLayer(conv.biases(), vecs);
+  DenormLayer(conv.weights(), vecs);
+}
+
+FloatVectors LoadFloatsFromPbFile(const std::string& filename) {
+  auto buffer = DecompressGzip(filename);
+  auto net = lc0::Net();
+  FloatVectors vecs;
+  net.ParseFromString(buffer);
+
```

Should throw exception as unsupported if net.format is not linear16.